source/retryable-reads/tests/README.rst

```diff
@@ -0,0 +1,269 @@
+      one or more of the following fields:
+
+      - ``error``: Optional. If ``true``, the test should expect an error or
+        exception. Implicitly true if ``result`` contains ``errorContains``,
+        ``errorCodeName``, ``errorLabelsContain``, and/or
+        ``errorLabelsOmit``. If ``false``, the test should expect no error or
+        exception, even if ``result`` contains ``errorContains``,
+        ``errorCodeName``, ``errorLabelsContain``, and/or ``errorLabelsOmit``.
+
+      - ``errorContains``: Optional. A substring of the expected error message.
+
+      - ``errorCodeName``: Optional. The expected "codeName" field in the server
+        error response.
+
+      - ``errorLabelsContain``: Optional. A list of error label strings that the
+        error is expected to have.
+
+      - ``errorLabelsOmit``: Optional. A list of error label strings that the
+        error is expected not to have.
+
+  - ``expectations``: Optional list of command-started events.
+
+MapReduce Tests
+===============
+
+Drivers MUST ensure that ``mapReduce`` is retried only when ``out: { inline:
+1}``.  The tests for ``mapReduce`` should test against the same conditions as
+described in ``aggregate.yml`` and ``aggregate-serverErrors.yml``, with the
+exception that ``mapReduce`` MUST not retry when the `out` stage is not
+``{inline: 1}``.
+
+A recommended data set would be three documents that contain values for x such
+that ``x:0``, ``x:1`` and ``x:2``. The map function should increment each of
+these values by one, and the reduce function should add all x's together.
+    
+GridFS Tests
+============
+
+Since the GridFS API is implemented using ``find`` commands, the `File
+Download`_ and `Generic Find on File Collection`_ parts of the GridFS API should
+be retryable under the same conditions that ``find`` is retryable. These
+functions should be tested against the same conditions described in ``find.yml``
+and ``find-ServerErrors.yml``. The tests' setup and expectations can be adapted
+from the "Download by Name when revision is 0" test at
+https://github.com/mongodb/specifications/blob/master/source/gridfs/tests/download_by_name.yml
+and the "Download when there is one chunk" test at
+https://github.com/mongodb/specifications/blob/master/source/gridfs/tests/download.yml.
+
+.. _File Download: https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#file-download-by-filename
+
+.. _Generic Find on File Collection:  https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#generic-find-on-files-collection
+
+Optionally Retryable Commands
+=============================
+
+If a driver chooses to implement retryability for the optionally retryable
+enumeration commands (e.g. ``client.listDatabases()``,
+``db.listCollections()``), the tests for the enumeration commands should test
+against the same conditions as described in ``find.yml`` and
+``find-serverErrors.yml``.
+
+    
+Replica Set Failover Test
```

Comment:
Do I assume correctly that this is pending SERVER-34943, per the note above about not being able to test with multi-node replicasets?

Reply:
Correct! I'll add another note about the need for that ticket to be resolved.